### DISCUSSION TITLE
More hooks

### DISCUSSION THREAD

**Discussion by sammy-ette** [sammy-ette/Hilbish#51] [Ideas]
in hilbish 0.2 the `bait` module was added to throw and catch hooks. the hooks added were `command.success` and `command.fail`
in hilbish 0.3 those 2 hooks were changed to a single `command.exit`
but there's no other official hilbish hooks in existence. want one to be added/have a good idea? send comment pls im out of ideas thank u

**Comment by sammy-ette** [comment]
`command.not-found`: as the name suggests, this will be specifically thrown when a user attempts to run a command that isnt found in the $PATH

**Reply by sammy-ette** [reply]
done

**Comment by sammy-ette** [comment]
~~`debug.error`: another pretty obvious one, debug errors will be thrown here. so if a user wants hilbish debug logs, they can use this~~
~~other `debug` hooks can be: `debug.info`, `debug.warn`~~

**Comment by nerd190** [comment]
Hello @TorchedSammy 👋
**Apologies for the off-topic bit!**
I've been following this project for a couple months now, VERY impressive!!!!!
I love shells, but no new shells have have got me excited quite like this one has, mainly because they all focus on the wrong thing (like nushell, that focuses on pretty printing 🤦‍♂️), hilbish is very different! after reading the discussions (like native package management etc) it feels very much like hilbish is doing the neovim thing, just taking what we already have and making it a bit better, not completely "re-inventing the wheel", obviously the biggest thing "out-of-the-norm" here is using Go and Lua, and I personally LOVE this!!!
Go's goroutines/multi-core/multi-threading/async support (something shells are missing) along with Lua's simple syntax and extra abilities (coroutines, tables etc) would make for an AMAZING shell!
I was a little worried when readline was picked as thats the only reason I use zsh! but hilbiline sounds promising (I understand that whilst zle is advanced, its a huge hack!)
So basically, well-done! and thank you!
**Back on-topic! HOOKS!**
One thing that always bothered me is zsh's lack of hooks, tmux uses hooks well! vim also.
I believe that shells are lacking "job" based hooks, that can report to us about any long-running commands, whether its foreground or background, something like:
```lua
job.did()    -- yay! :)
job.fail()    -- nay. :(
```
Or even just `job.done` and that reports info?

This can be done on zsh, but not natively, there are tons of plugins, but I think info (mainly errors) about my jobs should be native.
This way, if my long-running compile job in the background fails, I can either send the error to a desktop notifcation or my email or my mobile... etc. Bliss! 😊

Some shells have directory-changed hooks already but none have vim's `textYanked` hook, this would also be great! e.g to keep a list of yanks in a file that I can fuzzy-find and paste from (I do this in zsh, but not natively, I have to map `y` & `Y` to a zle widget I wrote that also passes the yanks to a file) this would also be great!

P.S As this project grows, also consider making a Matrix chat room for general stuff? I don't want to spam your Github! 😂

**THANK YOU FOR HILBISH!**

**Reply by sammy-ette** [reply]
that's a long post i'd never expect.
first thanks for your interest in hilbish! i don't really know what else to say.
hilbish development has been on hiatus for a month or 2 but im getting back into it now, and for that i should prioritize hilbiline for proper windows support and some niceties for our shell only (prompt with an updating clock?)

for the hook suggestion i'm just going to consider `job.done` for the jobs. `textYanked` seems a bit specific but can you explain it a bit more (and the `job.done` hook as well, like if it should pass anything or whatever)

> also consider making a Matrix chat room for general stuff?

i don't use matrix, and i already have a personal discord server. though if hilbish *does* get popular enough to be of use for a couple users ill consider a separate discord server and/or irc channel

**Comment by nerd190** [comment]
> i don't use matrix, and i already have a personal discord server

Discord is good for me! :)

> going to consider `job.done`

**Thank you!!!**

> can you explain it a bit more?

Of course!
I was thinking it could return three strings:
- The command (e.g `configure && make && make install &`)
- Its exit status (e.g `127`)
- Its error message (e.g `install: cannot create directory ‘/usr’: Read-only file system`

The hook should be configurable, else what is a long-running command? 11 seconds? 30 seconds? what do we mask? I dontt consider commands like `vi|vim|nvim|tmux|fzf|lf` but others may want `emacs|screen|skim|ranger`, zsh uses `zstyle` to configure these things without polluting the environment with variables for its builtins (perhaps this idea could be adopted by hilbish?)
The hook could have a default behavior i.e print to stdout, but obviously the user could redirect this wherever they want with a lua script, away from pc? send a text to me, or a discord channel, or simply a log either all my commands or just the errors (`${XDG_CACHE_HOME:-$HOME/.cache}/hilbish/errors.log`) .... 
Of course, this all takes zsh's `precmd` and `preexec` hooks.

> `textYanked` seems a bit specific.

Shells lack registers, we cut/copy/paste alot in the shell but you can only retrieve your last paste (easily), unless you use zsh's vi-mode, you can retrieve the last ten yanks, just like in vim. Problem? how do you see these yanks? where are they? `vim` has `:registers` allowing us to view them and plugins like `fzf.vim` can allow us to interactively find and select a yank, rather than trying to remember that `$MY_SECRET` is the 9th yank. Its another reason why I use zsh, I can define my own `yank` or `paste` or `cut` widget (function) and I can script my own keybindings for `Alt-P` which brings up `fzf` that filters my yanks so I can interactively select one (just like in vim), or a plugin could script emac's `killring` implementation (user pastes the first register, then `Alt-P`/`Alt-Shift-P` rotates through them), zsh can achieve both of these without this hook, by allowing us to define our own zle widgets, zle is powerful, but without that, the only way would be hooks to view  yanks, something like `text.yanked()`.

**Almost forgot!**
Something like `dir.changed()` or `wd.changed()` (directory changed).

**Reply by sammy-ette** [reply]
hm
with the explanation of `textYanked` it seems that's gonna have to wait until hilbiline is usable.
i'll see if i can do the job hook.

> Something like dir.changed() or wd.changed() (directory changed).

`cd` hook already exists, fun fact

late reply, school stuff plus needed a tiny break

**Reply by sammy-ette** [reply]
basically everything here has been added now (even textYanked, which is hilbish.vimAction; check the docs with `doc vim-mode action` command)
